numpy/_core/src/multiarray/mapping.c

```diff
@@ -1176,6 +1176,10 @@ array_assign_boolean_subscript(PyArrayObject *self,
     }
 
     v_data = PyArray_DATA(v);
+    if (v_stride == 0) {
+        /* Scalar/broadcast values benefit from KEEPORDER to match self layout */
+        order = NPY_KEEPORDER;
```

Comment:
Agreed that this is good enough, but might as well move it into the `v_stride` branch above?
Since I nitpicky, I might say that the comment could say: If the same value is repeated, iteration order does not matter.